### ISSUE TITLE
pageReset not working when my table completes an update

### ISSUE THREAD

**Issue by pashih** [Mottie/tablesorter#975]
I'm using tablesorter 2.22.3 with the pager plugin. I can't seem to get my table to reset to the first page when it completes updating. I'm not using the filter widget, could that be the problem? If so, can you guide me on what I need to change in order to reset my table to the first page when no parameters other than 'page' gets updated in the ajax url I send to my server?

**Comment by Mottie** [comment]
Hi @pashih!

Are you seeing the request being sent to the server? Have you tried setting the `debug` option to `true`?

Normally the pager plugin attempts to prevent sending duplicate requests to the server. So, it should be sending a request to the server if the page changes. You can bypass this and *force* a request by triggering the [`pagerUpdate` method](http://mottie.github.io/tablesorter/docs/#pagerupdate).

It would make it easier to figure out the problem if you could share the tablesorter & pager initialization code. The pager ajax options might be where the issue lies.

**Comment by pashih** [comment]
Hi @Mottie,

I'll give the debug option a shot. In the meantime, here's the pager init code (prepareRequestParams is a function that preps my custom filters and returns the serialized query string):
$('.tablesorter-bootstrap').tablesorter({
        theme : "bootstrap",
        headerTemplate : '{content}',
        widgets : ["uitheme", "zebra"],
        widthFixed : true
    }).tablesorterPager({
        container: $(".pager"),
        ajaxUrl: "/matchmakers/ajax.php?type=updateExploreUsers&page={page}&size={size}",
        customAjaxUrl: function(table, url) {
            $('tbody').html('Loading...');
            var params = prepareRequestParams();
            url += '&' + params;
            $(table).trigger('changingUrl', url);
            return url;
        },
        ajaxProcessing: function(ajax) {
            if(ajax && ajax.hasOwnProperty('data')) {
                return [ajax.total_rows, ajax.data];
            }
        },
        output: '{startRow}-{endRow} of {totalRows}',
        updateArrows: true,
        pageReset: 0,
        page: 0,
        size: 75,
        cssNext: '.next',
        cssPrev: '.prev',
        cssFirst: '.first',
        cssLast: '.last',
        cssPageDisplay: '.page-display',
        cssDisabled: 'disabled',
        cssGoto: '.goto-page'
    }).bind('pagerComplete', function() {
        if($('tbody').html() == 'Loading...') {
            $('tbody').html('');
        }
    });

**Comment by pashih** [comment]
Turned on the debug option but nothing seems to have changed. Are the pager settings somehow cached on my browser? If so, do I need to somehow force the new settings to be captured?

**Comment by pashih** [comment]
Actually looking into it some more... it looks like the pager settings are shared across different browser tabs. Ie. I'm on page 7 on one tab. When I load the same table on another tab it loads page 7 as well. Is that expected behavior? Is there a way to change that?

**Comment by Mottie** [comment]
Hmm, ok. I think I'm confused. When I read this:

> I can't seem to get my table to reset to the first page when it completes updating

I wonder how is the table being updated? Are the pager arrows working? What about the page select dropdown? Or, how else is the page being set?

When the user sets a page, the table should be loading the results for that page and not the first.

As for the tabs thing, the pager shouldn't do that unless both tables have the same ID, are in the same group, or have the same `wo.storage_tableId` option setting ([ref](http://mottie.github.io/tablesorter/docs/index.html#function-storage)). You can always set the [`savePages` option](http://mottie.github.io/tablesorter/docs/index.html#save-pages) to `false` so the last pager size & page is not saved and restored.

**Comment by pashih** [comment]
Here's a visual for the controls that I use for my table. The pager controls (arrows and dropdown) do work properly to select my pages. I use the customAjaxUrl option to attach my custom filter selections to the ajax call. I trigger the pagerUpdate event whenever my custom filters are changed.

The main issue I'm running into is that when my custom filters change and I trigger the table update (via pagerUpdate), the pager is still stuck on the same page it's currently on (and not reset to the first page as set by the pageReset option).

I'll give the savePages option a try. 

![screen shot 2015-07-24 at 12 48 40 pm](https://cloud.githubusercontent.com/assets/915872/8883351/f4eb0fd6-3202-11e5-8837-e83336d33e37.png)

**Comment by pashih** [comment]
Actually I read the documentation and the code again and I think I kind of understand where I went wrong. Originally I thought pageReset would be used upon any update to the table but it looks like you don't use it if the filters the table knows about aren't changed (because it would then be a pager triggered update using the same set of filters). Since I don't use the filter widget or bind my custom filters to the table, it probably doesn't know my changes are actually filter-triggered changes (since I call pagerUpdate to update my table) and thus doesn't reset my page back to the first.

I ended up going with the work around of having changes on my custom filters trigger a click on '.first' which is the first page button on the pager when that button is not disabled (already on page 1) and trigger pagerUpdate otherwise.

if($('.first').hasClass('disabled')) {
        $('#users-table').trigger('pagerUpdate');
    }
    else {
        $('.first').trigger('click');
    }

**Comment by Mottie** [comment]
Ahh, so maybe it is because you're not using the filter widget. Actually, more precisely, it is because the `filterStart` event isn't being triggered so the pager [`pageReset` option](http://mottie.github.io/tablesorter/docs/index.html#pager-page-reset) is never checked, nor updated.

What you're going to have to do is modify the `customAjaxUrl` function:

```js
customAjaxUrl: function(table, url) {
  $('tbody').html('Loading...');
  var params = prepareRequestParams();
  url = url.replace( '&page=' + table.config.pager.page + '&', '&page=0&' ) + '&' + params;
  // $(table).trigger('changingUrl', url); // are you using this code?
  return url;
},
```

I commented out the `changeUrl` trigger because that was added as an example to show that other stuff can be done in the callback. Remove it, unless you really are using it for something.
